python/cudf/cudf/core/tools/datetimes.py

```diff
@@ -652,3 +658,220 @@ def _isin_datetimelike(
+                '2026-04-23 08:00:00'],
+                dtype='datetime64[ns]')
+
+    """
+    if tz is not None:
+        raise NotImplementedError("tz is currently unsupported.")
+
+    if normalize:
+        raise NotImplementedError(
+            "normalizing timestamps to midnight is unsupported."
+        )
+
+    if closed is not None:
+        raise NotImplementedError("closed is currently unsupported.")
+
+    if [start, end, periods, freq].count(None) > 1:
+        raise ValueError(
+            "Of the four parameters: start, end, periods, and freq, exactly "
+            "three must be specified"
+        )
+
+    dtype = np.dtype("<M8[ns]")
+
+    _periods_not_specified = False
+
+    if freq is not None:
+        if isinstance(freq, DateOffset):
+            offset = freq
+        else:
+            raise TypeError("`freq` must be a cudf.DateOffset object.")
+
+        if _check_mixed_freqeuency(offset):
+            raise NotImplementedError(
+                "Mixing fixed and non-fixed frequency offset is unsupported."
+            )
+        offset = offset.to_pandas()
+
+        if start is None:
+            end = cudf.Scalar(end, dtype=dtype)
+            start = cudf.Scalar(
+                pd.Timestamp(end.value) - (periods - 1) * offset, dtype=dtype
+            )
+        elif end is None:
+            start = cudf.Scalar(start, dtype=dtype)
+        elif periods is None:
+            start = cudf.Scalar(start, dtype=dtype)
+            end = cudf.Scalar(end, dtype=dtype)
+            periods = math.ceil(
+                int(end - start) / _offset_to_nanoseconds_lower_bound(offset)
+            )
+            _periods_not_specified = True
+
+            if periods < 0:
+                # end < start, return empty column
+                periods = 0
+            elif periods == 0:
+                # end == start, return exactly 1 timestamp (start)
+                periods = 1
+    else:
+        start = (cudf.Scalar(start, dtype=dtype).value.astype("int64"),)
+        end = (cudf.Scalar(end, dtype=dtype).value.astype("int64"),)
+        arr = cp.linspace(start=start, stop=end, num=periods,)
+        result = cudf.core.column.as_column(arr).astype("datetime64[ns]")
+        return cudf.DatetimeIndex._from_data({None: result})
+
+    if normalize:
+        old_dtype = start.value.dtype
+        start = cudf.Scalar(
+            start.value.astype("datetime64[D]").astype(old_dtype)
+        )
+
+    if "months" in offset.kwds or "years" in offset.kwds:
+        res = libcudf.datetime.date_range(start.device_value, periods, offset)
+    else:
+        end = (pd.Timestamp(start.value) + (periods - 1) * offset).to_numpy()
+        arr = cp.linspace(
+            start=start.value.astype("int64"),
+            stop=end.astype("int64"),
+            num=periods,
+        )
+        res = cudf.core.column.as_column(arr).astype("datetime64[ns]")
+
+    if _periods_not_specified:
+        res = res[res <= end]
```

Comment:
I am having trouble following this case - would you mind just helping me understand with an example?

Reply:
Sure - definitely should better document this logic here.

Under a specific case, when only `start`, `end` dates and `freq` is specified, we need to compute the number of elements in the sequence. For example:
```python
start='2020-01-15', end='2030-05-10', freq=cudf.DateOffset(years=1, months=5)
```
We need to pass these information to libcudf supporting function `calendrical_months_sequence`, which only accepts arguments: `size`, `init`, `months` (step). The first one `size` is the crux here: how many elements are there between `2020-01-15` to `2030-05-10` with an interval of 17 months? 

We can get the timedelta in nanoseconds between these two dates by subtracting one from another. But we don't know what's `17 months` in nanoseconds because it's a non-fixed frequency. What we chose to do here is to make a pessimistic estimate using the lower bound of nanoseconds that 17 months could possibly be. Since each months has at least 28 days, it has at least `28*3600*1e9` nanoseconds. With this we can compute the upper bound of the `size` parameter. But the result will (possibly) contain results that's out of bound, so we need to perform a trimming at the end.